### ISSUE TITLE
BenchmarkResultCreate schema: rename, add constraints, remove constraints

### ISSUE THREAD

**Issue by jgehrcke** [conbench/conbench#823]
This defines the structure of the `stats` property in a BenchmarkResult object. Tweak name of the schema to better reflect that.

`times` is required and documented as

> If `data` is a duration measure, this should be a duplicate of that object

This is not enforced. Passing an emtpy list is what clients do. Maybe do not require this. But if it's passed, validate against `data`. If it's passed, maybe also require the values to be different from `data`.

The `iterations` field is required. Currently specified as

> Number of iterations that were executed (should be the length of `data` and `times`)

What if the value does not match the length? Maybe remove this parameter altogether.

**Comment by jgehrcke** [comment]
related: https://github.com/conbench/conbench/pull/416#discussion_r1009553017

**Comment by alistaire47** [comment]
> The `iterations` field is required. Currently specified as
> 
> > Number of iterations that were executed (should be the length of `data` and `times`)
> 
> What if the value does not match the length? Maybe remove this parameter altogether.

I think there may be some code somewhere that treats `iterations` as the intended number of iterations, but only populates `data` with successful iterations, omitting failures. This is worse than passing nulls (you can't tell which iteration[s] failed), so we should check implementations and update any relying on this. We should be able to query postgres and see if any like that have been submitted, though we're also going to get a bunch of malformed microbenchmarks that confuse iterations and replications (which we'll also need to fix if we're still submitting any like that).

**Comment by alistaire47** [comment]
> `times` is required and documented as
> 
> > If `data` is a duration measure, this should be a duplicate of that object
> 
> This is not enforced. Passing an emtpy list is what clients do. Maybe do not require this. But if it's passed, validate against `data`. If it's passed, maybe also require the values to be different from `data`.

💯 on board with making `times` optional; it's very confusing why its there when the unit for `data` is `"s"` already. `times` presently only makes sense for microbenchmarks where the unit for `data` is bytes/s or iterations/s.

More generally, we should do some initial thinking about multimetrics and how we can make this part of the schema more modular so we can pass things like memory usage along with timings and whatever. Currently `times` is kind of a second less-flexible metric, but not a very good one.

**Comment by jonkeane** [comment]
## `data` versus `times`

What @alistaire47 brought up at the end is spot on. The `data`/`times` schism was something of a shim to allow for a single extra metric (that we don't do much with inside of conbench). Like Ed said: The feature that they enable is for  benchmarks that use throughput as their measure, the `times` field lets us also store + save time alongside that. 

We might be able to rip out times entirely (though we should work with the folks we know who use it it now, Arrow C++ microbenchmarks + Go microbenchmarks to confirm that this won't hurt them too much if we do, of course), but personally I would rather leave that oddity as is for now and spend any time + effort adjusting it on doing multi-metrics properly (even if that means we have a slightly inconsistent or slightly weird API for a little while). We've already gotten (multiple) requests for multi-metrics. And AFAIU, proper multi-metrics support will involve touching this exact code + add code to deal with multiple metrics in a modular way, but also won't be an order of magnitude different amount of work + coordination compared to removing `times` now. But I'm willing to be convinced that this oddity really is getting in our way + we really absolutely do need to remove it in order to unblock us from moving on with other features (in which case let's commit + get it done).

## `iterations`, `data`, and `times` consistency

As for the iterations number + the lengths of those `data` and `times` lists: I'm strongly in favor of documenting and enforcing those shapes to be how we want them (more than they already are — I will admit I don't have off the top of my head all of the places they might already be documenter or enforced!).

As for the actual consistencies | behaviors I would expect: 

* `data` and `times` should be at a minimum the same length. If they are not, we should reject.
* `iterations` IMO should be allowed to be different from the length of `data`/`times` (if we keep it around). If we want those two to be enforced to be the same, we should get rid of `iterations` entirely and calculate it from the length of the list, which honestly might be the best way to go. Having them be separable from each other allows for some flexibility when sending microbenchmarks where you don't have the individual data points, but you do have summary stats + number of iterations run (example below), but honestly this is not used by anyone as far as I know, so we shouldn't do too much to (half) support it. But _ideally_ we would have a flexible enough setup that doing something like that was possible. 


Using something like `bench::mark()`, we could populate our stats with the min, median, etc. from there directly without sending the whole list of iteration times (but when also include `iterations = 10000`). 

```
> bench::mark(c(1:10))
# A tibble: 1 × 13
  expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result     memory             time                gc                   
  <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>     <list>             <list>              <list>               
1 c(1:10)       295ns    414ns  1798898.        0B        0 10000     0     5.56ms <int [10]> <Rprofmem [0 × 3]> <bench_tm [10,000]> <tibble [10,000 × 3]>
```
